main.ts

```diff
@@ -0,0 +1,381 @@
+        if (direction == RcbVerticalDirection.Up) {
+            rampUp(distance, horizSpace);
+        }
+        else {
+            rampDown(distance, horizSpace);
+        }
+    }
+
+    function rampUp(height: number, horizSpace: number) {
+        let unpoweredBlocksPlaced = 8; // Set to 8 so first block is powered.
+        for (let currentHeight = 0; currentHeight <= height; currentHeight++) {
+            for (let currentHoriz = 0; currentHoriz < horizSpace; currentHoriz++) {
+                if (unpoweredBlocksPlaced >= 8) {
+                    rollerCoasterBuilder.addPoweredRail()
+                    unpoweredBlocksPlaced = 0
+                } else {
+                    addUnpoweredPoweredRail()
+                    unpoweredBlocksPlaced++
+                }
+                builder.move(FORWARD, 1)
+            }
+            builder.move(UP, 1);
+        }
+        builder.move(DOWN, 1)
+    }
+
+    function rampDown(descentDistance: number, horizSpace: number) {
+        for (let currentDescent = 0; currentDescent <= descentDistance; currentDescent++) {
+            for (let currentHoriz = 0; currentHoriz < horizSpace; currentHoriz++) {
+                // Place powered at start only if needed, then every powerInterval blocks.
+                // Only needed on first descent level since the rest have the downhill to speed up.
+                let powerAtStart = currentDescent == 0 && horizSpace >= powerInterval;
+                if ((currentHoriz + (powerAtStart ? 0 : 1)) % powerInterval == 0) {
+                    rollerCoasterBuilder.addPoweredRail()
+                }
+                else {
+                    rollerCoasterBuilder.addRail()
+                }
+                builder.move(FORWARD, 1)
+            }
+            builder.move(DOWN, 1)
+        }
+
+        // Undo the final down movement, since we didn't actually place a block.
+        builder.move(UP, 1)
+    }
+
+    //% block="add $direction turn"
+    //% blockId="rcbAddTurn" weight=85
+    export function addTurn(direction: TurnDirection) {
+        rollerCoasterBuilder.addRail();
+        builder.move(FORWARD, 1);
+        rollerCoasterBuilder.addRail();
+        builder.turn(direction);
+        builder.move(FORWARD, 1);
+        rollerCoasterBuilder.addRail();
+        builder.move(FORWARD, 1);
+    }
+
+    //% block="add spiral going $verticalDirection turning $turnDirection with width $width and height $height"
+    //% width.min=3 width.defl=3
+    //% height.min=1 height.defl=10
+    //% blockId="rcbAddSpiral" weight=80
+    export function addSpiral(verticalDirection: RcbVerticalDirection, turnDirection: TurnDirection, height: number = 10, width: number = 3) {
+        let totalHeightDiff = 0
+        while (totalHeightDiff < height) {
+            let heightChange = verticalDirection == RcbVerticalDirection.Up && totalHeightDiff == 0 ? width - 1 : width - 2
```

Comment:
Should it be `height -1` and `height - 2` rather than `width`?

Also, use `===` for the comparisons.

Reply:
We actually do want width here. The width of the spiral determines how much we can change the height by in a single "stretch" (i.e. space between turns). So we change height by as much as possible based on the width, then turn, then change height by as much as possible again, turn, and so on... so the width is what determines the height change on each iteration of the loop.